File: 1653-minimum-deletions-to-make-string-balanced/1653-minimum-deletions-to-make-string-balanced.py
class Solution:
    def minimumDeletions(self, s: str) -> int:
        stack = list()
        result = 0

        for c in s[::-1]:
            if not stack or stack[-1] >= c:
                stack.append(c)
            else:
                stack.pop()
                result += 1

        return result

File: 1653-minimum-deletions-to-make-string-balanced/1653-minimum-deletions-to-make-string-balanced.java
class Solution {
    public int minimumDeletions(String s) {
        int[] dp = new int[s.length() + 1];
        int cnt = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == 'a') {
                dp[i + 1] = Math.min(dp[i] + 1, cnt);
            } else {
                dp[i + 1] = dp[i];
                cnt++;
            }
        }

        return dp[s.length()];
    }
}